packages/voting-contract/src/lib.rs

```diff
@@ -197,18 +198,24 @@ where
 {
     // anyone can trigger this if the vote passed
 
-    let mut prop = proposals::<P>().load(deps.storage, proposal_id)?;
-    if [Status::Executed, Status::Rejected, Status::Passed]
+    let mut prop = proposals().load(deps.storage, proposal_id)?;
+
+    if prop.status == Status::Rejected {
+        return Err(ContractError::Rejected {});
+    }
+
+    prop.update_status(&env.block);
```

Comment:
> > I think the idea with this stuff is that queries use current_status() and executions use update_status() and then everything checks out.
> 
> Yeah, you're right. I just realized, that despite status in memory not being correct, in the end it is correct for end user who uses query. We "simply" must be carefull to make sure we use that helper everywhere.

Just wrote a variant of the same above.

Reply:
Same here. If the code errors below, this status update will be discarded. And if it does not, the status will be saved later, so, there's really no need to call `update_status` instead of `current_status`.

The bug is, not calling any of those, and reading the status directly.